Clean up Segwit Recovery feature (D3194)

sickpig opened the pull request:
This is a port of ABC [D3194](https://reviews.bitcoinabc.org/D3194)   ("Clean up Segwit Recovery feature"):

- Invert meaning of Segwit Recovery flag.
- Remove segwit recovery activation
- Make SCRIPT_DISALLOW_SEGWIT_RECOVERY a standard flag

It also port [D3416](https://reviews.bitcoinabc.org/D3416) ("clean up script_tests -- move segwit recovery into static json") 

This based on @dagurval's #1881

Mengerian committed:
Clean up Segwit Recovery feature

Invert meaning of Segwit Recovery flag.
Remove segwit recovery activation
Make SCRIPT_DISALLOW_SEGWIT_RECOVERY a standard flag

To test it did an IBD on main net and testnet with `checkpoints=0`.
Update functional test to check for accepting blocks containing segwit recovery transactions,
rejecting segwit recovery transactions from mempool acceptance, and correct network banning.

This is a port of ABC https://reviews.bitcoinabc.org/D3194

markblundeberg committed:
clean up script_tests -- move segwit recovery into static json

Summary:
The `script_build` generated tests in script_tests are best reserved for
those tests that have transaction signatures in them -- they may need to
be regenerated for some reason as they are sensitive to transaction
contents and hashtype. Though non-signature tests can be done this way,
they unnecessarily add to the (quite large) computational resources used
during script_tests.cpp compilation and cause a warning during build:
`variable tracking size limit exceeded with -fvar-tracking-assignments, retrying without`

Since segwit recovery tests involve no signatures, they can be moved into
the static parts of the json file.

Thisis a port of ABC: https://reviews.bitcoinabc.org/D3416

dagurval commented on qa/rpc-tests/segwit-recovery.py:
```diff
@@ -234,91 +268,41 @@ def check_for_no_ban_on_rejected_tx(tx, reject_code, reject_reason):
+        # Blocks containing segwit spending txns are accepted in both nodes.
         block(5)
-        postforkblock = update_block(5, [txspend])
+        postforkblock = update_block(5, [txspend, txspend_case0])
```
Nit: `postforkblock` doesn't make sense, there is no forking.

dagurval commented on src/txadmission.cpp:
```diff
@@ -654,12 +654,8 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
-            0;
 
-    const uint32_t featureFlags = cds_flag | schnorrflag | segwit_flag;
+    const uint32_t featureFlags = cds_flag | schnorrflag;
```
It looks like the logic here changed from _allowing segwit recovery when non-standard transactions are accepted into mempool_, to _never allow segwit recovery transactions to mempool_.

This is also tested in the QA test.

Just pointing this out in case this was not intentional.

sickpig replied:
AFAIU the aim of the D3194 is to accept historical blocks that contains SegWit recovery txns, whereas at the same time such txns won't be accepted in the mempool.

dagurval replied:
As was the old behavior, but additionally it now also rejects segwit recovery transactions from mempool when `acceptnonstdtxn=1` is set.

I think this new behavior is fine. Allowing non-standard transactions to be mined should not imply allowing segwitted coins to be stolen.

sickpig replied:
Yes, agreed. with this changes there's no way to get segwit recovery txs transaction accepted in your mempool.

gandrewstone replied:
I disagree with this change.  If the feature was added in consensus we should support it.  What is probably going on is that certain miners are being provided a custom release in the hopes that only they can recover segwit transactions.  We should not make a change that would deny our miners revenue unless they jump thru hoops.

sickpig replied:
@gandrewstone 
so you are proposing to accept segwit recovery txns in the mempool in case `acceptnonstdtxn=1`?

I've no problem since either way we are going to accept blocks accepting blocks that contains this kind of transaction, so there's no risk of chain split in case we diverge from ABC in that particular regard.

What do you think @dagurval?

dagurval replied:
It's more correct and clean to keep old behavior, but I don't like it.

sickpig replied:
Ok I'm going to remove this code change so that we could still accept segwit recovery transaction by setting `acceptnonstdtxn=1`

sickpig replied:
done

dagurval commented:
LGTM

sickpig committed:
remove unused variable

markblundeberg committed:
remove ComparisonTestFramework dependency from segwit recovery test

This is a port of ABC https://reviews.bitcoinabc.org/D3674

sickpig committed:
remove another unused variable

jasonbcox committed:
Fix incorrect node being checked in segwit recovery test

Discovered while working on backports related to send_txs_and_test.
Before this patch, the test happens to pass because the check at line
270 is looking for a tx not being in the mempool of node0, whe

This is a port of ABC  https://reviews.bitcoinabc.org/D3859

sickpig committed:
Accept SegWit recovery transactions if acceptnonstdtxn=1

* remove segwit disallow flag from standard set in case acceptnonstdtxn=1
* add description string for SCRIPT_ERR_CLEANSTACK
* qa test still need to be fixed

gandrewstone commented:
having a flag in the negative like "DISALLOW_SEGWIT_RECOVERY" bends engineer's brains, especially when there's lots of different flags.  however we have a tradeoff between minimizing consensus code differences with ABC to make merges easier (which is the whole point of this PR) or being cleaner.  going with minimizing differences on this one...
